src/python/pants/backend/python/util_rules/lockfile_diff.py

```diff
@@ -0,0 +1,117 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import itertools
+import logging
+from dataclasses import dataclass
+from typing import Any, Mapping
+
+from packaging.version import LegacyVersion, Version, parse
+
+from pants.backend.python.util_rules.pex_requirements import (
+    LoadedLockfile,
+    LoadedLockfileRequest,
+    Lockfile,
+    LockfileContent,
+)
+from pants.core.goals.generate_lockfiles import (
+    LockfileGenerateDiff,
+    LockfileGenerateDiffResult,
+    LockfileRequirements,
+    RequirementName,
+)
+from pants.engine.fs import Digest, DigestContents
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.engine.unions import UnionRule
+
+logger = logging.getLogger(__name__)
+
+
+class PythonLockfileGenerateDiff(LockfileGenerateDiff):
+    pass
+
+
+@dataclass(frozen=True, order=True)
+class PythonRequirementVersion:
+    _parsed: Version | LegacyVersion
+
+    @classmethod
+    def parse(cls, version: str) -> PythonRequirementVersion:
+        return cls(parse(version))
+
+    def __str__(self) -> str:
+        return str(self._parsed)
+
+    def __getattr__(self, key: str) -> Any:
+        return getattr(self._parsed, key)
+
+
+def _pex_lockfile_requirements(lockfile_data: Mapping[str, Any] | None) -> LockfileRequirements:
+    if not lockfile_data:
+        return LockfileRequirements({})
+
+    # Setup generators
+    locked_resolves = (
+        (
+            (RequirementName(r["project_name"]), PythonRequirementVersion.parse(r["version"]))
+            for r in resolve["locked_requirements"]
+        )
+        for resolve in lockfile_data["locked_resolves"]
+    )
+
+    try:
+        requirements = dict(itertools.chain.from_iterable(locked_resolves))
+    except KeyError as e:
+        from pprint import pformat
+
+        logger.debug(f"Failed to parse PEX lockfile: {e}\n{pformat(lockfile_data)}")
+        requirements = {}
+
+    return LockfileRequirements(requirements)
+
+
+@rule
+async def generate_python_lockfile_diff(
+    request: PythonLockfileGenerateDiff,
+) -> LockfileGenerateDiffResult:
+    lockfile = request.lockfile
+    new_content = await Get(DigestContents, Digest, lockfile.digest)
+    old, new = await MultiGet(
+        Get(
+            LoadedLockfile,
+            LoadedLockfileRequest(
+                Lockfile(
+                    file_path=lockfile.path,
+                    file_path_description_of_origin="generated lockfile",
+                    resolve_name=lockfile.resolve_name,
+                ),
+                parse_lockfile=True,
+            ),
+        ),
```

Comment:
This `Get` may fail in case this is the first time this lockfile is generated in which case there will not be an existing file to glob up from the workspace, due to the glob match error behaviour of:

https://github.com/pantsbuild/pants/blob/fd8a9df06dcf2586e94571b924889a6a91840eb3/src/python/pants/backend/python/util_rules/pex_requirements.py#L169-L176

Hence, I'd like to use #17745 to guard against this case here, rather than turning the rule for `LoadedLockfile` into a version that supports `MaybeLoadedLockfile`, that would be more code bloat for this edge case...

Reply:
Solved by #17911